lib/vector-core/buffers/src/disk_v2/ledger.rs

```diff
@@ -0,0 +1,334 @@
+    pub fn increment_reader_file_id(&self) {
+        self.reader_current_data_file_id
+            .fetch_add(1, Ordering::AcqRel);
+    }
+}
+
+pub struct Ledger {
+    // Path to the data directory.
+    data_dir: PathBuf,
+    // Advisory lock for this buffer directory.
+    ledger_lock: LockFile,
+    // Ledger state.
+    state: BackedArchive<MmapMut, LedgerState>,
+    // Notifier for reader-related progress.
+    reader_notify: Notify,
+    // Notifier for writer-related progress.
+    writer_notify: Notify,
+    // Last flush of all unflushed files: ledger, data file, etc.
+    last_flush: AtomicCell<Instant>,
+    // How often flushes should occur.
+    //
+    // Flushes may occur more often as a data file filling up forcefully triggers a flush so that
+    // all data is on-disk before moving on to the next data file.
+    flush_interval: Duration,
+}
+
+impl Ledger {
+    pub fn state(&self) -> &ArchivedLedgerState {
+        self.state.get_archive_ref()
+    }
+
+    pub fn get_current_reader_data_file_path(&self) -> PathBuf {
+        self.get_data_file_path(self.state().get_current_reader_file_id())
+    }
+
+    pub fn get_current_writer_data_file_path(&self) -> PathBuf {
+        self.get_data_file_path(self.state().get_current_writer_file_id())
+    }
+
+    pub fn get_next_writer_data_file_path(&self) -> PathBuf {
+        self.get_data_file_path(self.state().get_next_writer_file_id())
+    }
+
+    pub fn get_data_file_path(&self, file_id: u16) -> PathBuf {
+        self.data_dir.join(format!("buffer-data-{}.dat", file_id))
+    }
+
+    /// Waits for a signal from the reader that an entire data file has been read and subsequently deleted.
+    pub async fn wait_for_reader(&self) {
+        self.reader_notify.notified().await;
+    }
+
+    /// Waits for a signal from the writer that data has been written to a data file, or that a new
+    /// data file has been created.
+    pub async fn wait_for_writer(&self) {
+        self.writer_notify.notified().await;
+    }
+
+    /// Notifies all tasks waiting on progress by the reader.
+    pub fn notify_reader_waiters(&self) {
+        self.reader_notify.notify_waiters();
+    }
+
+    /// Notifies all tasks waiting on progress by the writer.
+    pub fn notify_writer_waiters(&self) {
+        self.writer_notify.notify_waiters();
+    }
+
+    /// Determines whether or not all files should be flushed/fsync'd to disk.
+    ///
+    /// In the case of concurrent callers when the flush deadline has been exceeded, only one caller
+    /// will get a return value of `true`, and the others will receive `false`.  The caller that
+    /// receives `true` is responsible for flushing the necessary files.
+    pub fn should_flush(&self) -> bool {
+        let last_flush = self.last_flush.load();
+        if last_flush.elapsed() > self.flush_interval {
+            if self
+                .last_flush
+                .compare_exchange(last_flush, Instant::now())
```

Comment:
The semantics are a little different -- you have to loop -- but consider using `compare_exchange_weak`. The code generated on ARM will be much better.

Reply:
Yeah, I'm not really 100% happy with this code, and my thought was that I would just end up with a soft flush and hard flush, and only drive the hard flush (basically the `fsyncdata` syscall) via an external task/interval, but that would definitely not be super misuse-resistant... hmmmmmmmm.

Either way, yeah, I can change this to the weak variant for now.

Reply:
I actually went and checked: `crossbeam-utils` already uses `compare_exchange_weak` under the hood, so we should be gucci here. 🚀